consensus/istanbul/proxy/proxy_handler.go

```diff
@@ -0,0 +1,352 @@
+
+	log.Info("Proxy handler stopped")
+	return nil
+}
+
+// isRunning returns if `run` is currently running in a goroutine
+func (ph *proxyHandler) isRunning() bool {
+	ph.lock.RLock()
+	defer ph.lock.RUnlock()
+
+	return ph.running
+}
+
+// run handles changes to proxies, validators, and performs occasional check-ins
+// that proxy/validator assignments are as expected
+func (ph *proxyHandler) run() {
+	defer ph.loopWG.Done()
+
+	phEpochTicker := time.NewTicker(ph.proxyHandlerEpochLength)
+	defer phEpochTicker.Stop()
+
+	logger := ph.logger.New("func", "run")
+
+	ph.updateValidators()
+
+loop:
+	for {
+		select {
+		case <-ph.quit:
+			// The proxyHandler was stopped
+			break loop
+
+		case addProxyNodes := <-ph.addProxies:
+			// Got command to add proxy nodes.
+			// Add any unseen proxies to the proxy set and add p2p static connections to them.
+			for _, proxyNode := range addProxyNodes {
+				proxyID := proxyNode.InternalNode.ID()
+				if ph.ps.getProxy(proxyID) != nil {
+					logger.Debug("Proxy is already in the proxy set", "proxyNode", proxyNode, "proxyID", proxyID, "chan", "addProxies")
+					continue
+				}
+				log.Info("Adding proxy node", "proxyNode", proxyNode, "proxyID", proxyID)
+				ph.ps.addProxy(proxyNode)
+				ph.sb.AddPeer(proxyNode.InternalNode, p2p.ProxyPurpose)
+			}
+
+		case rmProxyNodes := <-ph.removeProxies:
+			// Got command to remove proxy nodes.
+			// Remove the proxy and remove the p2p static connection
+			for _, proxyNode := range rmProxyNodes {
+				proxyID := proxyNode.ID()
+				proxy := ph.ps.getProxy(proxyID)
+				if proxy == nil {
+					logger.Warn("Proxy is not in the proxy set", "proxy", proxyNode, "proxyID", proxyID, "chan", "removeProxies")
+					continue
+				}
+
+				logger.Info("Removing proxy node", "proxy", proxy.String(), "chan", "removeProxies")
+
+				if valsReassigned := ph.ps.removeProxy(proxyID); valsReassigned {
+					ph.sendValEnodeShareMsgs()
+					ph.sb.UpdateAnnounceVersion()
+				}
+				ph.sb.RemovePeer(proxy.node, p2p.ProxyPurpose)
+			}
+
+		case connectedPeer := <-ph.addProxyPeer:
+			// Proxied peer just connected.
+			// Set the corresponding proxyInfo's peer
+			peerNode := connectedPeer.Node()
+			peerID := peerNode.ID()
+			proxy := ph.ps.getProxy(peerID)
+			if proxy != nil {
+				logger.Debug("Connected proxy", "proxy", proxy.String(), "chan", "addProxyPeer")
+				if valsReassigned := ph.ps.setProxyPeer(peerID, connectedPeer); valsReassigned {
+					logger.Info("Remote validator to proxy assignment has changed.  Sending val enode share messages and updating announce version")
+					ph.sendValEnodeShareMsgs()
+					ph.sb.UpdateAnnounceVersion()
+				}
+
+				// Share this node's enodeCertificate for the proxy to use for handshakes via a forward message
+				proxyEnodeCertificateMsg, err := ph.sb.GenerateEnodeCertificateMsg(proxy.externalNode.URLv4())
+				if err != nil {
+					logger.Warn("Error generating enode certificate message", "err", err)
+				} else {
```

Comment:
nit: here and below. use `continue` instead of `if..else` nesting to handle errors